File: app/src/main/java/com/mdgd/pokemon/ui/pokemons/dto/PokemonsScreenState.java
package com.mdgd.pokemon.ui.pokemons.dto;

import com.mdgd.pokemon.models.repo.schemas.PokemonDetails;

import java.util.LinkedList;
import java.util.List;

public class PokemonsScreenState {

    private static final int SET_DATA = 1;
    private static final int ADD_DATA = 2;
    private static final int ERROR = 3;
    private static final int LOADING = 4;
    private static final int UPDATE_DATA = 5;

    private final List<PokemonDetails> list;
    private final int action;
    private final Throwable error;

    public PokemonsScreenState(int action) {
        this.action = action;
        this.list = null;
        this.error = null;
    }

    public PokemonsScreenState(int action, List<PokemonDetails> list) {
        this.action = action;
        this.list = list;
        this.error = null;
    }

    public PokemonsScreenState(int action, Throwable error) {
        this.action = action;
        this.list = new LinkedList<>();
        this.error = error;
    }

    public static PokemonsScreenState createSetDataState(List<PokemonDetails> list) {
        return new PokemonsScreenState(SET_DATA, list);
    }

    public static PokemonsScreenState createUpdateDataState(List<PokemonDetails> list) {
        return new PokemonsScreenState(UPDATE_DATA, list);
    }

    public static PokemonsScreenState createAddDataState(List<PokemonDetails> list) {
        return new PokemonsScreenState(ADD_DATA, list);
    }

    public static PokemonsScreenState createErrorState(Throwable error) {
        return new PokemonsScreenState(ERROR, error);
    }

    public static PokemonsScreenState createLoadingState() {
        return new PokemonsScreenState(LOADING);
    }

    public void visit(PokemonsScreen pokemonsScreen) {
        if (LOADING == action) {
            pokemonsScreen.showProgress();
            return;
        }
        pokemonsScreen.hideProgress();
        if (SET_DATA == action) {
            pokemonsScreen.setItems(list);
        } else if (ADD_DATA == action) {
            pokemonsScreen.addItems(list);
        } else if (UPDATE_DATA == action) {
            pokemonsScreen.updateItems(list);
        } else if (ERROR == action) {
            pokemonsScreen.showError(error);
        }
    }
}

File: app/src/main/java/com/mdgd/pokemon/ui/pokemons/PokemonsScreenState.kt
package com.mdgd.pokemon.ui.pokemons

import com.mdgd.mvi.ScreenState
import com.mdgd.pokemon.models.repo.dao.schemas.PokemonFullDataSchema

sealed class PokemonsScreenState(protected val list: MutableList<PokemonFullDataSchema>,
                                 protected val availableFilters: List<String>,
                                 protected val activeFilters: MutableList<String>) : ScreenState<PokemonsContract.View> {
    protected var isHandled = false

    fun getItems(): List<PokemonFullDataSchema> {
        return ArrayList(list)
    }

    override fun visit(screen: PokemonsContract.View) {
        screen.setItems(list)
        for (filter in availableFilters) {
            screen.updateFilterButtons(activeFilters.contains(filter), filter)
        }
    }

    @JvmName("getActiveFilters1")
    fun getActiveFilters(): List<String> {
        return ArrayList(activeFilters)
    }


    class Initial(availableFilters: List<String>)
        : PokemonsScreenState(ArrayList(0), availableFilters, ArrayList(0))

    class Loading(lastState: PokemonsScreenState)
        : PokemonsScreenState(ArrayList(lastState.list), lastState.availableFilters, ArrayList(lastState.activeFilters)) {

        override fun visit(screen: PokemonsContract.View) {
            super.visit(screen)
            if (!isHandled) {
                screen.showProgress()
                isHandled = true
            }
        }
    }

    class SetData(list: List<PokemonFullDataSchema>, lastState: PokemonsScreenState)
        : PokemonsScreenState(ArrayList(list), lastState.availableFilters, ArrayList(lastState.activeFilters)) {

        override fun visit(screen: PokemonsContract.View) {
            screen.hideProgress()
            super.visit(screen)
        }
    }

    class AddData(list: List<PokemonFullDataSchema>, lastState: PokemonsScreenState)
        : PokemonsScreenState(ArrayList(lastState.list), lastState.availableFilters, ArrayList(lastState.activeFilters)) {

        init {
            super.list.addAll(list)
        }

        override fun visit(screen: PokemonsContract.View) {
            screen.hideProgress()
            super.visit(screen)
        }
    }

    class UpdateData(items: List<PokemonFullDataSchema>, lastState: PokemonsScreenState)
        : PokemonsScreenState(ArrayList(items), lastState.availableFilters, ArrayList(lastState.activeFilters)) {

        override fun visit(screen: PokemonsContract.View) {
            screen.hideProgress()
            screen.updateItems(list)
        }
    }

    class Error(val error: Throwable?, lastState: PokemonsScreenState)
        : PokemonsScreenState(lastState.list, lastState.availableFilters, ArrayList(lastState.activeFilters)) {

        override fun visit(screen: PokemonsContract.View) {
            screen.hideProgress()
            super.visit(screen)

            if (!isHandled) {
                screen.showError(error)
                isHandled = true
            }
        }
    }

    class ShowDetails(val id: Long?, lastState: PokemonsScreenState)
        : PokemonsScreenState(lastState.list, lastState.availableFilters, ArrayList(lastState.activeFilters)) {

        override fun visit(screen: PokemonsContract.View) {
            screen.hideProgress()
            super.visit(screen)

            if (!isHandled) {
                screen.proceedToNextScreen(id)
                isHandled = true
            }
        }
    }

    class ChangeFilterState(filter: String, lastState: PokemonsScreenState)
        : PokemonsScreenState(ArrayList(lastState.list), lastState.availableFilters, ArrayList(lastState.activeFilters)) {

        init {
            if (activeFilters.contains(filter)) {
                activeFilters.remove(filter)
            } else {
                activeFilters.add(filter)
            }
        }

        override fun visit(screen: PokemonsContract.View) {
            super.visit(screen)
            screen.hideProgress()
        }
    }
}
